### DISCUSSION TITLE
Version 4.3.0.rc1

### DISCUSSION THREAD

**Discussion by damianlegawiec** [spree/spree#11232] [General]
# Modular API-first architecture

**Monolith is gone replaced with a truly headless modular design!**

We've completely decoupled Storefront (`spree_frontend`), Admin Panel (`spree_backend`) and extracted transactional Emails into their own gem (`spree_emails`). Thanks to this you can:

   1. Easily replace Storefront and Admin Panel with your own
   2. Use the version of Storefront and Admin you like, so you can update the Spree Core and API without the need to upgrade these two
   3. Easily replace standard transactional emails with Klaviyo, Mailchimp and other 3rd parties
 
We're also making Spree less dependent on the Rails framework while supporting Rails 5.2, 6.0 and 6.1 on Ruby 2 and 3.

# Pages & Navigation CMS

This release gives Spree users a powerful content management system that goes beyond creating basic text pages.

Spree Content Management System allows you to create:

**Home Pages**
- Create many home pages per store.
- Each home page is unique by language.
- Add multiple sections to each home page.

**Standard Pages**
- Standard pages consist of WYSIWYG text editor content.
- Each standard page has a slug that is unique to the store.
- Each page is assigned a language.

**Feature pages**
- Feature pages use page sections.
- Each feature page has a slug that is unique to the store.
- Feature pages are assigned a language.

**Menus**
Users can create navigation menus for their Spree stores. 

Each menu is unique by language and location within the store and has the following features:
- Menus contain many menu items. Each menu item can link to a product, taxon, URLs, home page or CMS page.
- Menu items can be nested inside other menu items creating nested links.
- Menus can have container items used to organize links within the menu.
- Menu items can have images allowing you to create promotions within the main menu.

# Platform API Developer Preview

A new set of fast and modern APIs for application to application integrations based on oAuth 2.0 authentication and permissions.
This API completely replaces API v1 which will be extracted from Spree core in 5.0. Platform API is built on the same  foundations and technology as [Storefront API](https://api.spreecommerce.org/docs/api-v2/api/docs/v2/storefront/index.yaml) that means JSON API format, Swagger (Open API) documentation and oAuth 2.0 authentication.

[Platform API Documentation](https://api.spreecommerce.org/docs/api-v2/YXBpOjg3MzkxODk-platform-api-v2)

Platform API is the biggest and most robust API for Spree with the biggest number of API endpoints to cover all features available in Admin Panel plus more. All API v1 usage in Admin Panel was also replaced with Platform API.

# Advanced Product Filters

We've greatly expanded the product filters feature. Now you can:

1. Filter by Properties - you can set which Properties should be filterable and the ones you choose will be visible and supported in the Storefront and API
2. Filter by Option Types - you can set which Option Types should be filterable and the ones you choose will be visible and supported in the Storefront and API
3. Filters on Storefront, both Properties and Option Types will be automatically shown only when Products in selected Taxon have these Option Types/Properties attached, same goes for values, we're only displaying values from Products in the given Taxon

This way you can create new custom filters in the Admin UI from Option Types or Properties and the system will do the rest work for you.

# Installation

1. [Install Docker](https://docs.docker.com/get-docker/)
2. Download [Spree Starter](https://github.com/spree/spree_starter/archive/main.zip)
3. Unzip it
4. Run `bin/setup` in `spree_starter-main` directory

# Upgrade

[Follow 4.2 to 4.3 upgrade guide](https://dev-docs.spreecommerce.org/upgrades/upgrades/fout-dot-two-to-four-dot-three)

# Other noteworthy changes

Please review each of the noteworthy changes to ensure your customizations or extensions are not affected. If you are affected by a change and have any suggestions, please submit a PR to help the next person!

## API

* Added caching to API v2 serialized increasing API responsiveness 3-5 times [Damian Legawiec](https://github.com/spree/spree/pull/10875)

    This also includes new confirmation option `Spree::Api::Config[:api_v2_cache_ttl]` for the cache expiration TTL. The default value is `3600` (1 hour). The cache also auto-expires when the cached record is updated, more on this topic: https://github.com/jsonapi-serializer/jsonapi-serializer#caching
    
* **Multi-Store** Storefront Products API returns Products from the current Store [Damian Legawiec](https://github.com/spree/spree/pull/11126)
* **Multi-Store** Storefront Account Orders API returns Orders from the current Store [Damian Legawiec](https://github.com/spree/spree/pull/11126)
* **Multi-Store** Storefront Order Status API returns Order from the current Store [Damian Legawiec](https://github.com/spree/spree/pull/11126)

## Core

* **Multi-Store** `Order#available_payment_methods` by default will return only Payment Methods available in Order's Store [Damian Legawiec](https://github.com/spree/spree/pull/11126/commits/8f52301c8178e04bb1aa6a03cde5ebb9f0063cbb)

    Passing `store` argument to that method will result in deprecation warning
* **Multi-Store** Deprecated `Store.current` in favour of ` Stores::FindCurrent` [Damian Legawiec](https://github.com/spree/spree/pull/11126/commits/f3414d67b92a2b1d2eb920abab95ff48ab8afd72)

  Also, this finder class can be replaced by custom one by setting 
  
  ```ruby
  Spree::Dependencies.current_store_finder = YourCustomStoreFinder
  ``` 
  in `config/initializers/spree.rb` (please check [documentation](https://guides.spreecommerce.org/developer/customization/dependencies.html))
* **Multi-Store** `Product`, `Promotion` and `PaymentMethod` models require at least one `Store` associated [Damian Legawiec](https://github.com/spree/spree/pull/11141)
* **Multi-Store** `Promotion` is now associated to multiple `Store` records via `StorePromotion` model [Damian Legawiec](https://github.com/spree/spree/pull/11167)
    
    You can disable this behaviour by setting:

    ```ruby
    Spree::Config[:disable_store_presence_validation] = true
    ```

    in your `config/initializers/spree.rb` file

* **Multi-Store** `Store Credit` model is now associated with `Store` [Szymon Iwacz](https://github.com/spree/spree/pull/11155)
* `Taxon` model now requires `Taxonomy` presence and that `Taxonymy` needs to match Taxon's Parent (if present) [Damian Legawiec](https://github.com/spree/spree/pull/11214)
* Transactional emails were extracted into a separate gem called `spree_emails` [Damian Legawiec](https://github.com/spree/spree/pull/11110/commits/1ebb0dd8c7f8f7e296f21c630886b05869b6acbc)
* Replaced `twitter_cldr` gem with much smaller `validates_zipcode` [Damian Kaczmarczyk](https://github.com/spree/spree/pull/11077)
* Removed `premailer-rails` gem dependency [Damian Legawiec](https://github.com/spree/spree/pull/11041/commits/2726da8db3992306fcde58175e17a379029e2955)
* Removed `rails` dependency - only require Rails gems that we really use [Damian Legawiec](https://github.com/spree/spree/pull/11041/commits/2dc16c59204257e7e01cee386d9cad4824f4c107)
* Removed `responders` dependency [Damian Legawiec](https://github.com/spree/spree/pull/11041/commits/9751b9b9cd02c3f88a7727eade676f6635c4da09)
* Removed `sprockets` dependency [Damian Legawiec](https://github.com/spree/spree/pull/11041/commits/f87c44fbd3b4a99c6fa04cf089655e84efb29a56)
* Moved `spree.js` file to frontend/backend [Damian Legawiec](https://github.com/spree/spree/pull/11110/commits/57d1bd9d9d44525c2cbac769726bc617e6e7e809)
* Moved `polyfill.min.js` and `fetch.umd.js` files to frontend [Damian Legawiec](https://github.com/spree/spree/pull/11110/commits/e0d2901665cbf56633e8d86b2a822b4f5b916a0b)
* Moved `cleave.js` file to frontend/backend [Damian Legawiec](https://github.com/spree/spree/pull/11110/commits/2dc959d3c0c75c3efe7494ec754c127cb8a392d4)
* Moved `jquery.payment.js` file to frontend [Damian Legawiec](https://github.com/spree/spree/pull/11110/commits/3605d678486312eefd725f4ee6918c93bd3215ca)

## Storefront

* Upgraded **Sprockets** to **v4** and added support for **ES6**, **Source Maps** and **Manifest.js**  [Damian Legawiec](https://github.com/spree/spree/pull/10852)
* **Multi-Store** Only Store Orders are returned [Damian Legawiec](https://github.com/spree/spree/pull/11126/commits/a694445a99a41f36825666fe04f73f020951fa2e)
* **Multi-Store** Only Store Products are returned [Damian Legawiec](https://github.com/spree/spree/pull/11126/commits/e72ced9330e371211027d7a3371792d8fd5ed6e1)
* Added clearing all filters on PLP / Taxon page [Lukasz Adamczyk](https://github.com/spree/spree/pull/11031)
* Added filtering Products on PLP by a custom price range [Lukasz Adamczyk](https://github.com/spree/spree/pull/11007)
* Display only these Product Filters that have Products attached [Lukasz Adamczyk](https://github.com/spree/spree/pull/11023/files)

## Full Changelog

You can view the full changes using [Github Compare](https://github.com/spree/spree/compare/4-2-stable...4-3-stable).

<hr /><em>This discussion was created from the release <a href='https://github.com/spree/spree/releases/tag/v4.3.0.rc1'>Version 4.3.0.rc1</a>.</em>

**Comment by obaviet** [comment]
How do i setup domain for Multi store? can i use domain and sub domain?

**Reply by damianlegawiec** [reply]
@obaviet you can use both - you can set up domains via Admin UI -> Configurations -> Stores. DNS settings need to be set up at your DNS provider.

**Comment by JTimings** [comment]
@damianlegawiec what’s the ETA for 4-3-stable?

**Reply by damianlegawiec** [reply]
@JTimings EOD today :)

**Reply by damianlegawiec** [reply]
@JTimings it's here :) https://github.com/spree/spree/discussions/11339
